Bluetooth: host: Reserve conn object

joerchan opened the pull request:
Handle connection objects:
Reserve conn object for connectable advertiser. This means we won't have
a situation where we start a connectable advertise but will fail to
allocate a connection object for it in the connection complete event.

Reserve a connection object when starting the auto-initiator using the
controller whitelist.

Fixes:
Make sure that the auto-conn state is cleared correctly when we might
fail to allocate a new connection object.

Fix conn object assigned to the wrong controller connection in the
connection complete handler. This could happen when running a
directed advertiser and establishing a connection at the same time to
the same peer.

Refactor:
Refactor stopping directed advertiser to disconnect the state object
when calling advertise stop. This follows the same pattern as
bt_conn_disconnect.

Behaviour change:
Removed returning connection object already connected or in the process of establishing connection when using API to create new connection.
Removed returning disconnected object when creating new connection, this creates a consistent API behavior for when there are no free connection object when starting any Bluetooth LE roles that might result in a new connection being established.

Document:
Document the recommendation of handling no free connection objects in the disconnected callback.
Recommend to either increase connection count, or submit work to the system queue.

joerchan committed:
Bluetooth: host: Fix conn object assigned to wrong connection

Fix conn object assigned to the wrong controller connection in the
connection complete handler. This could happen when running a
directed advertiser and establishing a connection at the same time to
the same peer.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Clear auto-conn state when failed to get conn object

Make sure that the auto-conn state is cleared correctly when we might
fail to allocate a new connection object.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Reserve conn object for connecting with whitelist

Reserve a connection object when starting the auto-initiator using the
controller whitelist.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refactor stopping directed advertiser

Refactor stopping directed advertiser to disconnect the state object
when calling advertise stop. This follows the same pattern as
bt_conn_disconnect.

Remove returning conn objects in BT_CONN_CONNECT state, this state could
only  be an initiator starting a connection in the central role.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Reserve conn object for connectable advertiser

Reserve conn object for undirected connectable advertiser. This means we
won't have a situation where we start a connectable advertise but will
fail to allocate a connection object for it in the connection complete
event.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jhedberg This is a follow up to the discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/21714#discussion_r363298381

@cvinayak I think this fixes the issue you saw once where the connection objects were assigned to the wrong role.

@kapi-no I think this fixes the issue where the host has less connection objects that the controller. Can you confirm?

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2319,19 +2319,6 @@ struct bt_conn *bt_conn_create_slave_le(const bt_addr_le_t *peer,
-				return NULL;
-			}
-
-		case BT_CONN_CONNECT:
```
Why is the `BT_CONN_CONNECT` case removed?

joerchan replied:
The connect state should be for initiator, after calling le_create_conn command. Shouldn't be a valid state for an advertiser.

jhedberg replied:
Ok, this wasn't clear from the commit message.

joerchan replied:
Good point. I forgot to mention it.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5939,13 +5932,31 @@ int bt_le_adv_start_internal(const struct bt_le_adv_param *param,
 		}
+
+		if (param->options & BT_LE_ADV_OPT_CONNECTABLE) {
+			conn = bt_conn_add_le(param->id, BT_ADDR_LE_NONE);
```
Since you set the address to `BT_ADDR_LE_NONE` here, don't you need to update code somewhere else to set the real address once we have a connection?

joerchan replied:
The same code that handled writing the addresses when the connection object was created in the connection complete event will overwrite the addresses in the connection complete event (same for all types of connections).

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1702,11 +1713,19 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
+		} else if (old_state == BT_CONN_CONNECT_ADV) {
+			if (conn->err) {
+				notify_connected(conn);
+			}
```
Notifying the application in this case doesn't seem right since we haven't exposed the connection to the application at this point. It would not even have a real address assigned to it.

joerchan replied:
Yes, that is a good point. Removing connectable advertiser from here.

What do you think about BT_CONN_CONNECT_AUTO?
In the other PR (adv/scan addr fix) I notify the application that the auto conn has been stopped by RPA timeout.

With this change I now have another option and that is to restart the auto-conn, since the conn-params are in the connection object. But I'm not convinced it is the best option.
So it is either: 1. stop at RPA timeout. 2 toggle whitelist initiator at RPA timeout.
Possible: 3: add different timeout on auto-conn, same as direct-conn.
Thoughs?

jhedberg replied:
I didn't realise we stopped automatic connection creation due to RPA timeout. I don't think we should do that, i.e. we should restart whatever automatic connection requests were ongoing after RPA timeout.

Btw, did you notice that there are some Shippable failures?

joerchan replied:
Yes, fixed shippable, needed to consider advertiser without peripheral.

At the moment, if the initiator is running when RPA timeout happen, we will fail to update the random address. I'm trying to fix it here: #21714
My current approach (int that PR) is to just disable the whitelist initiator, after having first made sure that the RPA timeout is restartet, so that the whitelist initiator will run for a full RPA period.
The reason for this approach was that I didn't have a copy of the conn params.

With this PR I have the conn param as they are in the conn-object. So I guess we should probably go with the restart approach? I think it needs an additional state variable to handle race-conditions between app stopping the initiator in the middle of the RPA timeout. Which is why I'm currently wondering if the simple stop approach is better.

jhedberg replied:
The automatic connection establishment is something that's supposed to continue over a long period of time, in most cases as long as the device is paired, so stopping it at RPA timeout doesn't seem right. Even though it might increase implementation complexity it'll reduce it on the app side, which is what should take priority (in addition to avoiding surprises in the API usage).

Btw, are you expecting #21714 to be reviewed & merged independently of this one, do you consider them to be mutually exclusive, or what?

joerchan replied:
Agreed. Then we will toggle the whitelist initiator at RPA timeout.

I created #21714 first, but it currently depends on changes in this PR. So should be reviewed after this one has been merged and I have made the update to toggle the whitelist initiator.
I need the conn params to toggle the initiator, and the easiest solution to that is to use the conn params from the conn object, which is the dependency on this PR.

kapi-no commented:
@joerchan, yes, this is working. The peripheral doesn't advertise when the host runs out of all its memory slots for connection objects.

Previously, the host started to advertise when the controller had the capacity (memory) to do so. The host didn't take into account its own memory restrictions.

Thank you for the fix!

joerchan commented:
@jhedberg 

```
		notify_disconnected(conn);

		/* Release the reference we took for the very first
		 * state transition.
		 */
		bt_conn_unref(conn);
```
This is the reason we cannot start new advertiser or whitelist initiator in the disconnected callback.

One possebility could be this to document that if the application wants to use the conn object in the disconnected callback it needs to have one reference to it.
And then do this:
```
		/* Release the reference we took for the very first
		 * state transition.
		 */
		bt_conn_unref(conn);
		notify_disconnected(conn);
```

The application would then have to do this:
```
static void disconnected(struct bt_conn *conn, u8_t reason)
{
	char addr[BT_ADDR_LE_STR_LEN];
	int err;

	if (conn != default_conn) {
		return;
	}

	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));

	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);

	bt_conn_unref(default_conn);
	default_conn = NULL;

        bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
}
```

This way we could avoid re-using the conn object in the disconnected state.

joerchan commented:
Alternatively we could document a safe way for the application to restart it. I guess using
`k_delayed_work_submit(&start_adv_or_initiator, K_NO_WAIT);` would be the best approach.

joerchan commented:
Putting a DNM until we clarify the behaviour of starting the advertiser from the disconnected callback. Something that we currently do in our applications.

joerchan commented:
@carlescufi @jhedberg Comments on this: https://github.com/zephyrproject-rtos/zephyr/pull/21743#issuecomment-572026635?

jhedberg commented:
> bt_conn_unref(conn);
> notify_disconnected(conn);

@joerchan even if we'd be checking for refcount > 1 that'd be abusing the reference counting principles, so I'd prefer to exclude this option from consideration.

jhedberg commented:
`k_delayed_work_submit(&start_adv_or_initiator, K_NO_WAIT);` that's actually synonymous to `k_work_submit()`, but yes, I'd be fine with suggestion this option for code that desires reuse and can't afford to increase `CONFIG_BT_MAX_CONN`

carlescufi reviewed (commented):
@jhedberg @joerchan 

**Option 1**:

This option is probably the least evil, but it implies a change in the application that needs to be documented in the release notes, since the app can no longer be assured to get a valid `conn` object in the `disconnected` callback.

```
static void conn_update_timeout(struct k_work *work)
{
	if (conn->state == BT_CONN_DISCONNECTED) {
		bt_l2cap_disconnected(conn);
                 refc = conn->refcount;
                 bt_conn_unref(conn);
                if (refc > 1) {
		    notify_disconnected(conn);
                  }
                  else {
		    notify_disconnected(NULL);
                  }
		return;
	}
...

struct bt_conn *bt_conn_create_le(const bt_addr_le_t *peer,
				  const struct bt_le_conn_param *param)
{
 
//		case BT_CONN_DISCONNECTED:
//			BT_WARN("Found valid but disconnected conn object");
//			goto start_scan;
}

struct bt_conn *bt_conn_create_slave_le(const bt_addr_le_t *peer,
					const struct bt_le_adv_param *param)
{
//		case BT_CONN_DISCONNECTED:
//			BT_WARN("Found valid but disconnected conn object");
//			goto start_adv;
}
...
static void disconnected(struct bt_conn *conn, u8_t reason)
{
        if (conn) {
             /* Remove the application's reference to the connection */
	     bt_conn_unref(conn);
        }
        bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
}

static void disconnected(struct bt_conn *conn, u8_t reason)
{
        if (conn) {
             addr = bt_conn_get_dst(conn);
             /* Remove the application's reference to the connection */
	     bt_conn_unref(conn);
        }
        /* Won't be reused because we've unrefed it */
        bt_conn_create_le(addr);
}

static void disconnected(struct bt_conn *conn, u8_t reason)
{
        if (conn) {
             /* Remove the application's reference to the connection */
	     bt_conn_unref(conn);
        }
        bt_conn_create_auto_le();
}
```

**Option 2**:

Document that you need a `k_work_submit()` in order to create a new connection no matter which API you use:

- `bt_le_adv_start()` (connectable)
- `bt_conn_create_slave_le()`
- `bt_conn_create_le()` 
- `bt_conn_create_auto_le()` 

and remove the reusing of the conn:

```
struct bt_conn *bt_conn_create_le(const bt_addr_le_t *peer,
				  const struct bt_le_conn_param *param)
{
 
//		case BT_CONN_DISCONNECTED:
//			BT_WARN("Found valid but disconnected conn object");
//			goto start_scan;
}

struct bt_conn *bt_conn_create_slave_le(const bt_addr_le_t *peer,
					const struct bt_le_adv_param *param)
{
//		case BT_CONN_DISCONNECTED:
//			BT_WARN("Found valid but disconnected conn object");
//			goto start_adv;
}
```

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -6007,6 +6026,37 @@ int bt_le_adv_stop(void)
+	struct bt_conn *adv_conn = NULL;
+	int err;
+
+	if (IS_ENABLED(CONFIG_BT_PERIPHERAL) &&
```
Wont the linker strip it away from the final image if there are no callers of it?

joerchan replied:
If CONFIG_BT_CENTRAL is defined, then connected and disconnected HCI events will be included and there will be calls to it. But the condition will never be true if not CONFIG_BT_PERIPHERAL is also enabled. So I think carles' suggestion is correct.

joerchan replied:
Given the same argument why this must be under BT_PERIPHERAL, then this will only be called for connectable advertising. So changed to assert on this assumption, and removed all if(conn) in this function.

joerchan committed:
Bluetooth: host: Document connection object behaviour

Add documentation for why connection objects are still in use during the
disconnected callback and document error code when starting connectable
advertiser with no free connection object available.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix error code when calling API before init

Fix bt_conn_create_aute_le returning the wrong error code when bt_init
has not been called yet. This is inconsistent with the rest of the API
functions.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Stop using existing conn object when creating new conn

Remove re-using connection objects in disconnected state when creating
directed advertiser or establishing a connection as a central using
direct connection procedure.
This makes the API mores consistent it terms of which connection roles
can be started from the disconnected callback.
This also avoids a central connection object being re-used for a
connection as a peripheral instead and vice versa.

When attempting to create a new connection the API would returning
a valid connection object if there is already an existing connection
object.
This existing connection object could be either in the process of
establishing the connection or already connected.
Returning the connection object in this would give the false impression
that the stack has initiated connection procedure, when in fact it just
returned an existing connection object.

The application has the ability to check for existing connection objects
using the bt_conn_lookup_addr_le API.

Add warning plus comment possible scenarios why the a valid connection
object might exists. Most important is to explain why a valid connection
object exists during the disconnected callback.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jhedberg @carlescufi Updated to option 2 as discussed on slack. Also removed returning of the conn object in the connect or connected state (establishing or connected) as we agreed.

I did a search through the samples and we seem to be covered. Central samples usually start scanner.
Peripheral samples usually use the KEEP_ADVERTISING flag, this case is handled in this PR.
The following samples already used k_work_submit or k_delayed_work_submit:
- pong
- eddystone
- mesh_badge

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1695,8 +1712,22 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			}
 
 			bt_conn_unref(conn);
+		} else if (old_state == BT_CONN_CONNECT_AUTO) {
```
Doesn't have to be done in this PR, but this long list of tests and branches for `old_state` should probably be converted to a switch statement.

joerchan replied:
Made a note to change this outside of this PR.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2167,20 +2216,8 @@ struct bt_conn *bt_conn_create_le(const bt_addr_le_t *peer,
-			bt_conn_unref(conn);
-			return NULL;
-		}
+		bt_conn_unref(conn);
```
Maybe add a BT_WARN() here "existing connection object already found" or something like that.

joerchan replied:
Added warning printing the state string. Plus commented about the different states.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2291,38 +2327,15 @@ struct bt_conn *bt_conn_create_slave_le(const bt_addr_le_t *peer,
-			bt_conn_unref(conn);
-			return NULL;
-		}
+		bt_conn_unref(conn);
```
A BT_WARN() might be appropriate here, especially to give hints to out-of-tree apps that might have relied on a successful return previously.

joerchan replied:
Added.

joerchan replied:
Added.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5921,13 +5920,32 @@ int bt_le_adv_start_internal(const struct bt_le_adv_param *param,
 	err = set_advertise_enable(true);
 	if (err) {
+		BT_ERR("Failed to start advertiser");
+		if (IS_ENABLED(CONFIG_BT_PERIPHERAL) && conn) {
```
Since you're pre-initializing conn to NULL and only setting it to non-NULL in a branch that uses `IS_ENABLED(CONFIG_BT_PERIPHERAL)` I *think* you can drop the `IS_ENABLED()` here and the compiler should be smart enough to compile away the branch either way. It'll also make the code look safer since it won't raise the question "wait, are we leaking conn here if PERIPHERAL is not set but conn is set?".

joerchan replied:
Looks like you are right. mesh_demo passed. Will check if shippable finds other configurations

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5921,13 +5920,32 @@ int bt_le_adv_start_internal(const struct bt_le_adv_param *param,
 		return err;
 	}
 
+	if (IS_ENABLED(CONFIG_BT_PERIPHERAL) && conn) {
```
Same here - I suspect the `IS_ENABLED()` can be dropped

joerchan replied:
If you compare this with the reference handling of `bt_conn_create_le` (or slave) then this is the ref that would otherwise have been given to the application, and the application would have the responsebility of calling bt_conn_unref. This one balances the ref made by bt_conn_add_le.

The ref that is maintained by the stack is in the bt_conn_set_state(conn, BT_CONN_CONNECT_ADV);
This one will released when bt_conn_set_state(conn, BT_CONN_DISCONNECTED); transitions from a non-connected state.

To clarify
```
conn = bt_conn_add_le(param->id, BT_ADDR_LE_NONE); // Ref 1

if (!conn) {
	return -ENOMEM;
}

bt_conn_set_state(conn, BT_CONN_CONNECT_ADV); // Ref 2

err = set_advertise_enable(true);
if (err) {
	bt_conn_set_state(conn, BT_CONN_DISCONNECTED); // Ref 1
	bt_conn_unref(conn); // Ref 0
	return err; // Conn object freed on return
}

bt_conn_unref(conn); // Ref 1
return 0; // Conn object has valid reference, freed by bt_le_adv_stop, or disconnected callback


```

I can add this comment:
"Since we don't give the application a reference to manage in this case, we need to release this reference here."

joerchan replied:
Added comment for cases where we create the conn, but don't give it to the application to manage.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5980,6 +6014,37 @@ int bt_le_adv_stop(void)
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_PERIPHERAL) && adv_conn) {
```
Like in the other case where I commented on the same, I think the `IS_ENABLED()` could be removed here

joerchan replied:
This function is now undef #ifdef CONFIG_BT_PERIPHERAL so removed all IS_ENABLED here.

jhedberg reviewed (changes requested):
See my inline comments

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1088,20 +1088,26 @@ static void update_pending_id(struct bt_keys *keys, void *data)
-	if (conn) {
+	if (IS_ENABLED(CONFIG_BT_CENTRAL) && role == BT_HCI_ROLE_MASTER) {
+		conn = bt_conn_lookup_state_le(peer_addr, BT_CONN_CONNECT);
 		return conn;
```
This could be just `return bt_conn_lookup_state_le()`. Same for the other branch and then you can remove the conn variable completely from this helper function.

joerchan replied:
It might look wierd when looking at this commit and not on the overall diff. I intentionally made this commit this way to avoid the next commit rewriting it.
Next commit adds:
```
		if (!conn) {
			conn = bt_conn_lookup_state_le(BT_ADDR_LE_NONE,
						       BT_CONN_CONNECT_ADV);
		}

```
Any objections to doing that? I'd prefer to keep the commits separate since one is refactor, the other a change.

jhedberg replied:
I missed that part. In that case it's perfectly fine.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5964,6 +5989,21 @@ int bt_le_adv_stop(void)
 	}
 
+	if (IS_ENABLED(CONFIG_BT_PERIPHERAL)) {
+		conn = bt_conn_lookup_state_le(BT_ADDR_LE_NONE,
```
You can move the scope of the `conn` variable into this branch, rather than it being global to the function (following the principle of declaring variables in the smallest possible scope)

joerchan replied:
Yup. Moved it.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1265,6 +1265,12 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 	    IS_ENABLED(CONFIG_BT_WHITELIST) &&
 	    evt->role == BT_HCI_ROLE_MASTER) {
+		/*
+		 * Clear auto conn even if we are not able to add connection
```
Minor: please start the text on the first comment line, i.e. `/* Clean auto...`

I can see we haven't been completely consistent with this, which is probably reflected in the inconsistency in your own patches (here you do this, other comments you don't), however I'd prefer if we'd standardize on the starting with the first line. It's more compact, and (at least for me) it doesn't really hurt the readability.

joerchan replied:
I copy pasted the comment and edited it. Updated both places.

jhedberg reviewed (changes requested):
Looks generally good now, thanks! Just some few minor cosmetic things which I hope you still have the energy for and then we can merge and move on 😄

jhedberg commented:
Looks like CI is not happy because of this:
```
zephyr-eabi/bin/ld: zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(conn.c.obj): in function `conn_update_timeout':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/conn.c:348: undefined reference to `bt_le_adv_resume'
```

joerchan commented:
> Looks like CI is not happy because of this:
> 
> ```
> zephyr-eabi/bin/ld: zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(conn.c.obj): in function `conn_update_timeout':
> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/conn.c:348: undefined reference to `bt_le_adv_resume'
> ```

Yes, I forget the `bt_le_adv_resume` in when notifying with disconnected callback. It's an easy fix.

> Looks generally good now, thanks! Just some few minor cosmetic things which I hope you still have the energy for and then we can merge and move on.

Thanks. After checking and seeing the samples all use k_work_submit, and the fact that our downstream application also used k_work_submit. I feel much better about this.
The reason why we had a problem in the downstream application was because the disconnected callback was given before cleanup. So k_work_submit wasn't enough, we had to use k_delayed_work_submit with an uncertain delay. You changed it when you made the re-use commit,

Given this I think we made the right call today.

pdunaj commented:
@MarekPieta , can you have a quick look? It should work for us but just in case.
